[feat] 판매내역 조회 기능

pie2457 opened the pull request:
## Issues
- #36 

## What is this PR? 👓
판매내역 조회 기능에 대한 PR입니다.

## Key changes 🔑
- 패키지 구조를 변경했습니다.
- 판매내역 조회
  - 판매내역 전체 조회
  - `판매중 (+ 예약중)` 인 상품 판매내역 조회
  - `판매완료`인 상품 판매내역 조회
- 판매내역 조회 테스트
  - 판매내역 `전체` 조회 테스트
  - `판매중`인 상품 판매내역 조회 테스트
  - `판매완료`인 상품 판매내역 조회 테스트
- QueryDsl을 이용해 판매내역 조회 기능을 구현했습니다. 
## 📖 공부하면서 배웠던 것
### Enum의 비교는 ==일까 equals()일까 ?
![image](https://github.com/pie2457/Tomato-market/assets/104147789/96f6fd4d-523f-4958-ae7e-1102f1eab1b6)
enum의 equals()메서드를 먼저 살펴보니 내부에는 결국 == 연산자로 비교하고 있었다. 
즉, **enum 상수들끼리 비교할 때는 equals()로 비교하나 == 으로 비교하나 똑같다**는 뜻인데, 
내가 아는 ==은 동일성으로 두 개의 인스턴스가 주소값이 같다는 것이고,
equals()는 동등성으로 두 개의 인스턴스가 주소값은 다르지만 같은 value를 가지고 있는걸 의미했다.
분명 ==과 equals는 용도와 의미가 다른데 왜 같이 쓰는지 이해가가질 않아 찾아보니,
![image](https://github.com/pie2457/Tomato-market/assets/104147789/5dfc4262-f302-49e8-b47a-dd2af83bd329)
자바 공식 문서에 따르면 **enum 상수의 인스턴스는 딱 1개이기 때문에, 비교하려는 두 객체 중 적어도 하나가 enum의 상수라면 equals()대신 == 연산자로 비교할 수 있다**라고 되어있다! 
### 결론 : 싱글톤을 보장하는 enum이기 때문에 동일성과 동등성을 비교할 필요가 없게 된것이다.

pie2457 committed:
style #36 : 패키지 변경

pie2457 committed:
feat #36 : 판매내역 조회

pie2457 committed:
test #36 : 판매내역 조회 테스트

pie2457 committed:
refactor #36 : 판매내역 조회를 위해 RequestMapping 수정

pie2457 committed:
refactor #36 : itemId가 안나오는 문제 수정

response의 itemId를 바인딩 하는 과정에 이름을 잘못적어둬서 itemId를 response에 넣어주지 못함 : cursor -> itemId로 수정

23Yong commented on src/main/java/pie/tomato/tomatomarket/presentation/item/ItemController.java:
```diff
@@ -71,14 +72,23 @@ public ResponseEntity<Void> modifyItem(@PathVariable Long itemId,
 		return ResponseEntity.ok().body(itemService.itemDetails(itemId));
 	}
+
+	@GetMapping("/sales/history")
+	public ResponseEntity<CustomSlice<SalesItemDetailResponse>> salesItemDetails(
+		@AuthPrincipal Principal principal,
+		@RequestParam(required = false) String status,
+		@RequestParam(required = false, defaultValue = "10") int size,
+		@RequestParam(required = false) Long cursor) {
+		return ResponseEntity.ok().body(itemService.salesItemDetails(status, principal, size, cursor));
```
판매내역 조회 때문에 위의 `/items` 를 모두 없애는 것보다 SalesItemController 같은 컨트롤러 클래스를 하나 더 만드는 것도 좋다고 생각합니당

pie2457 replied:
그렇게 생각하긴 했는데 SalesItemController에는 지금 구현한 기능 하나밖에 없을꺼같은데 따로 분리하는게 나을까용? 만약 SalesItemController를 만든다면 SalesItemService도 있어야 할까요? 아니면 ItemService를 써도될까용

23Yong replied:
상품 테이블을 이용해서 ItemService를 그대로 사용해도 될 것 같아요~
정하기 나름인 것 같아요~!

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/item/ItemService.java:
```diff
@@ -185,4 +168,37 @@ private void verifyExistsItem(Long itemId) {
+
+	public CustomSlice<SalesItemDetailResponse> salesItemDetails(
+		String status, Principal principal, int size, Long cursor) {
+		if (status.equalsIgnoreCase("null")) {
+			new UnAuthorizedException(ErrorCode.NOT_LOGIN);
+		}
```
판매내역이 null 인 경우 모든 판매내역을 조회하는 것인데 
`new UnAuthorizedException(ErrorCode.NOT_LOGIN);`을 하는 이유가 있나용?
그리고 throw 키워드가 빠진 것 같네요~

pie2457 replied:
헉 throwㅋㅋㅋㅋㅋㅋㅋ 그러네요 하나가 빠져있네요ㅠㅠ
맨처음 들어오는 null은 로그인이 안된 사람이 판매내역 조회를 눌렀을때를 위함이고 모든 판매내역은 all로 정해놨습니다 !

23Yong replied:
로그인이 되어 있지 않은지는 `@AuthPrincipal` 애노테이션이 동작할 때 걸러질 것 같아요~

pie2457 replied:
으음 그렇네용 수정하겠습니당 !

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/item/ItemService.java:
```diff
@@ -185,4 +168,37 @@ private void verifyExistsItem(Long itemId) {
+		if (status.equalsIgnoreCase("null")) {
+			new UnAuthorizedException(ErrorCode.NOT_LOGIN);
+		}
+		ItemStatus findStatus = findStatus(status);
```
해당 로직을 `ItemStatus` 한테 넘겨도 좋을 것 같습니다.
```java
ItemStatus itemStatus = ItemStatus.from(status);
```

pie2457 replied:
from 메서드는 all을 검사할 수 없고 ItemStatus의 value는 한글로 "판매중" "판매완료"로 되어있어서 파라미터를 그대로 못받아 들여서 일부러 메서드를 따로 하나 만들어놨어요 !

23Yong replied:
그러면 value를 추가적으로 두고 아래와 같이 작성하는 것도 좋을 것 같네용
```java
Arrays.stream(ItemStatus.values())
    .filter(itemStatus -> itemStatus.name().equalsIgnoreCase(status))
    .findFirst()
    .orElse(null);
```
이렇게 작성하면 밑의 `findStatus`와 같은 동작을 할 것 같아요~

pie2457 replied:
넹 감사합니다 !

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/persistence/item/ItemRepository.java:
```diff
@@ -45,4 +46,14 @@ default BooleanExpression equalMemberId(Long memberId) {
 		return item.member.id.eq(memberId);
 	}
+
+	default BooleanExpression findStatus(ItemStatus status) {
+		if (status == null) {
+			return null;
+		} else if (status.equals(ItemStatus.SOLD_OUT)) {
+			return item.status.eq(status);
+		} else {
+			return item.status.ne(ItemStatus.SOLD_OUT);
+		}
```
if ~ else if ~ else 구문보다 아래와 같이 하는 것도 좋을 것 같아용

```java
if (status == ItemStatus.SOLD_OUT) {
    return itemStatus.eq(status);
}
return item.stauts.ne(ItemStatus.SOLD_OUT);
```

자바의 enum은 `==`으로 비교해도 된답니답. 그 이유를 아시나여?

pie2457 replied:
all이 들어왔을 경우 null 처리를 해서 null도 따로 뺴놨어요 ne는 판매중이랑 예약중만 포함되는거 같아서요 !
String 타입으로 인식하나요 ㅠ? 왜그런지 모르겠네요 찾아보겠습니다 ! 감사해요 !

23Yong commented on src/test/java/pie/tomato/tomatomarket/application/ItemServiceTest.java:
```diff
@@ -282,6 +283,55 @@ void itemDetails() {
+			"sold_out", principal, 10, null);
+
+		// then
+		assertThat(salesItemDetailResponse1.getContents().size()).isEqualTo(2);
+		assertThat(salesItemDetailResponse2.getContents().size()).isEqualTo(1);
```
여러 개의 검증(assert) 구문은 assertAll이나 softAssertion을 사용해서 묶을 수 있을 것 같네요~

pie2457 replied:
넹 assertAll로 수정할게요 ~!

23Yong reviewed (approved):
수고하셨어요~

pie2457 committed:
refactor #36 : 리뷰사항 반영

- ItemController, SalesItemController : 판매목록 조회 메서드를 따로 분리했다.
- ItemService : findStatus 메서드는 service의 영역이 아니라고 생각해 ItemStatus로 분리했다
- ItemServiceTest : item 인스턴스를 사용하지않아 변수를 삭제했다.
